Wrap around eslint-plugin-jsx-a11y/anchor-has-content to support createInterpolateElement usage

rowasc opened the pull request:
## What?

Fixes https://github.com/WordPress/gutenberg/issues/21441.

Changes:

- packages/eslint-plugin/configs/jsx-a11y.js: disables the rule 'jsx-a11y/anchor-has-content' in favor of wordpress/a11y-interpolated-anchor, which wraps around the former to allow both checking regular JSX and not failing on interpolated elements.

- a11y-interpolated-anchor.js adds the new wrapper rule. This rule is then enabled in the custom.js rules of the eslint-plugin.

## Why?

When using anchors and createInterpolateElement together, we need to eslint-ignore the line to avoid running into the anchor-has-content error even if the anchor is valid. 
This pull request introduces a wrapper around the rule so that we do not need to constantly ignore it.


## How?

- The new rule uses jsx-a11y/anchor-has-content under the hood to validate non-interpolated anchors (as normal). I considered just writing the rule from scratch, and I can still do that but as a first attempt, I decided it made sense to add a wrapper and not reinvent the linter rules for normal anchors.
- The new rule also does some very basic checks for the potential validity of anchor tags used within _createInterpolateElement_. In particular, it checks that anchors are not empty.
- The trade-off of having any linting at all for the interpolated anchor is that it could give a false sense of safety, but it would still miss issues (see the tests for some examples).


## Testing Instructions
<!-- Please include step by step instructions on how to test this PR. -->
<!-- 1. Open a Post or Page. -->
<!-- 2. Insert a Heading Block. -->
<!-- 3. etc. -->

### Testing Instructions for Keyboard

- "Break" an anchor that is created with createInterpolateElement, by making it empty of content (e.g: `<a> </a>`)
    - It should trigger a `@wordpress/a11y-anchor-has-content` linter error that says `Anchors must have content and the content must be accessible by a screen reader. Check the first parameter to createInterpolateElement` 
- Add some empty anchor with JSX like so `<a href=""></a></>`, without createInterpolateElement
    - It should trigger a `@wordpress/a11y-anchor-has-content` linter error that says `Anchors must have content and the content must be accessible by a screen reader`
- Run the tests for the eslint-plugin rules `npm run test:unit --  packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js`. The `a11y-anchor-has-content.js` tests should pass (and all others).

rowasc committed:
createInterpolateElement() no longer triggers jsx-a11y/anchor-has-content when using anchors

Changes:
- Disabled the rule 'anchor-has-content' in favor of wordpress/a11y-interpolated-anchor. Changes in: packages/eslint-plugin/configs/jsx-a11y.js
- Adds a new rule in a11y-interpolated-anchor.js and enables it under the custom.js rules in our eslint-plugin

Why: when using anchors and createInterpolateElement together, we need to eslint-ignore the line to avoid running into anchor-has-content. This commit tries to solve for this by wrapping around the rule.

Instead of using eslint-ignore for the jsx-a11y/anchor-has-content when we want to use createInterpolateElement and an anchor, we now wrap around the rule.

The new rule uses jsx-a11y/anchor-has-content under the hood to validate non-interpolated anchors (as normal)
The new rule also does some very basic checks for potential validity of anchor tags used within createInterpolateElement. In particular it checks that
- anchors are not empty
- anchors have an 'a' in the conversion map (but only cares about having one item)
- unmatched A tags in the string we're interpolating (e.g: 'Hello</a>' )

The trade-off of having any matchers for the interpolated anchor is that it could give a false sense of safety, but it may still miss issues. The tests will include a comprehensive list of cases that are covered.

rowasc committed:
Removed flaky rules

rowasc committed:
Add tests to validate key cases in a11y-interpolate-anchor

rowasc committed:
a11y-interpolated-anchor.js did not reflect the intent of the lint rule

rowasc committed:
Remove unused messageId since we no longer test for this

rowasc committed:
Improve explanations of what the errors and rule tests do to better reflect the difference between wrapper and eslint-plugin-jsx-a11y/anchor-has-content errors

rowasc committed:
Add usecase test for when we have a custom component in the anchor

rowasc commented on packages/eslint-plugin/configs/custom.js:
```diff
@@ -7,6 +7,7 @@ module.exports = {
 		'@wordpress/no-global-active-element': 'error',
 		'@wordpress/no-global-get-selection': 'error',
 		'@wordpress/no-unsafe-wp-apis': 'error',
+		'@wordpress/a11y-anchor-has-content': 'error',
```
I tried a few names, could not land on a better one but open to ideas. Decided on this just to emulate the original rule that we wrap around.

gziolo replied:
The original plugin name has also JSX in the name so maybe `@wordpress/jsx-a11y-anchor-has-content`?

rowasc replied:
Done! thanks for the suggestion

rowasc commented on packages/eslint-plugin/rules/__tests__/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,170 @@
+ */
+import rule from '../a11y-anchor-has-content';
+
+const ruleTester = new RuleTester( {
+	parserOptions: {
+		sourceType: 'module',
+		ecmaVersion: 6,
+		ecmaFeatures: {
+			jsx: true,
+		},
+	},
+} );
```
FWIW I mostly just used the examples from other rules, didn't really dig much into RuleTester. Flagging because maybe someone experienced with this can make suggestions on settings or anything I missed.

gziolo replied:
It’s fine to copy it. The `ecmaVersion` should get updated in all tests, but it's an unrelated issue. I think it's fine to copy it from another file.

rowasc commented on packages/eslint-plugin/rules/__tests__/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,170 @@
+		   };`,
+		},
+		{
+			// ⚠️ This is invalid code that would get flagged by jsx-a11y/anchor-is-valid, but NOT by eslint-plugin-jsx-a11y/anchor-has-content or @wordpress/a11-anchor-has-content.
```
I added these so that it's obvious to readers what's covered/not covered (also because I was afraid of us forgetting the scope of the linter otherwise), but LMK if they feel irrelevant or if you have suggestions on how to improve them.

gziolo replied:
It’s also possible to add names for tests according to https://eslint.org/docs/latest/integrate/nodejs-api#ruletester. Maybe that would be a way to provide additional context instead of using inline comments.

rowasc replied:
Thank you so much, completely missed the `name` field. I've switched to using name in most cases.

rowasc commented on packages/eslint-plugin/rules/__tests__/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,170 @@
+		   };`,
+		},
+		{
+			// ⚠️ This passes the linter, but it's actually invalid code that would fail in runtime with createInterpolateElement due to missing the conversion map.
```
Something I considered is just making the wrapper plainly ignore all createInterpolateElement + anchor usages, and create a separate linter or a separate context within this linter that only deals with createInterpolateElement.

But again, I went with a simpler version first to learn the basics of working w/ Gutenberg. LMK if you want me to go down that road.

rowasc commented on packages/eslint-plugin/rules/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,96 @@
+	schema: [], // no options
+};
+
+const getATagsContent = ( str ) => {
```
I could improve this with a tokenizer a la createInterpolate element and catch more complex cases (e.g unmatched tags) if we want to. Originally did that, but felt it was adding more complexity than needed and the lint rules ended up being a bit flaky (better to do a focused createInterpolateA11y type linter for that).

gziolo replied:
It’s better to start with something simpler and improve over time. The bigger the scope, the longer it takes to review and land. Everything is going to be better than disabling the rule everywhere the helper function is used.

rowasc committed:
Use node name. I can't remember why I used key.name originally and JSX nodes should have a name

rowasc commented on packages/eslint-plugin/configs/custom.js:
```diff
@@ -7,6 +7,7 @@ module.exports = {
 		'@wordpress/no-global-active-element': 'error',
```
⚠️ Missing: Docs! which I completely forgot until now. Leaving as draft until I get a little time do write some.

rowasc replied:
https://github.com/WordPress/gutenberg/pull/50065/files?short_path=e4b51f2#diff-e4b51f2bcad9f627a441e4941e91e50987e0ece0d4e4873675c46daaca18ccbe some initial docs here, they cover all basic scenarios

rowasc commented on packages/eslint-plugin/rules/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,95 @@
+			}
+			// createInterpolateElement's 1st argument (a string with markup).
+			const textParam = interpolatedEl.arguments[ 0 ] ?? [];
+
```
⚠️ Merge blocker
```suggestion
if ( textParam.type !== 'Literal' ) {
				return; // bail on identifiers, etc as done in valid-sprintf
			}
```

Either bail or implement a small shared library to get the value? since I think we ran into a similar situation in other linters (e.g valid-sprintf).

gziolo replied:
We can do it in steps (multiple PRs).

rowasc replied:
Changed this to handle the failing test from the previous version. Thank you!

bor0 commented on packages/eslint-plugin/rules/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,95 @@
+				return;
+			}
+			const interpolatedEl = context
+				.getAncestors()
```
How deep does this go in the tree? Can it potentially catch false positives in case there's a nested combination of `createInterpolateElement` and anchors?

rowasc replied:
It goes up to "Program" but there are no false positives as far as I can see. We only get one through `find`, and because the parent of an <a> will never be another <a>, and we don't navigate recursively through all the callees of each parent, I _think_ we are safe. 
See [this test](https://github.com/WordPress/gutenberg/pull/50065/files#diff-4c555e1042f88bdebf5cba43ba582ae78b721c7d64c1d0cc527ea1975ef191edR269-R278), where I tried to demonstrate the concept with 3 elements. LMK if you think I missed something!

bor0 replied:
Got it, thanks for double-checking and adding tests for the case!

> and because the parent of an will never be another

Per docs, it seems that `getAncestors` will actually return them in order, starting from the direct parent, so this should be good even if it returns all other entires (as `find` will match the first one, like you said)

gziolo commented on packages/eslint-plugin/rules/__tests__/a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,170 @@
+				},
+			],
+		},
+	],
```
So nice to see these tests added ❤️

rowasc replied:
Thanks! I've added some more to cover the scenarios I had missed (custom config, nested components) originally.

gziolo commented:
This is solid work, and this PR only needs some final touches to bring it to the finish line. Excellent job tackling this issue. I see that one of the CI jobs errors while running `npm run lint:js` so there needs to be and edge case in the implementation that unit tests don't catch:

![Wrap around eslint-plugin-jsx-a11yanchor-has-content to support createInterpolateElement usage · WordPressgutenberg@cdb2f9e](https://user-images.githubusercontent.com/699132/235258573-2d0bded3-7561-4999-97d3-fc0e4192fd56.png)

> Run the linter on some code that uses // eslint-disable-next-line jsx-a11y/anchor-has-content ( where the anchor has content )
> It should not have an jsx-a11y/anchor-has-content error.
> Delete the ignore rule // eslint-disable-next-line jsx-a11y/anchor-has-content in that code
> It should not have an jsx-a11y/anchor-has-content error nor a @wordpress/a11y-anchor-has-content

The old rule is disabled now, so it should be safe to remove all those comments from the code. Do I miss something here?

gziolo commented:
@ciampo and @tyxla, you might be interested in helping to bring this PR to the finish line as it will improve the linting of WordPress components and have a tiny impact on having better a11y support that is about to be audited in https://github.com/WordPress/gutenberg/issues/50166.

tyxla commented:
Sure thing, thanks for the ping @gziolo!

@rowasc how can we help?

getdave committed:
Mark related selector as resolved (#50324)

rowasc committed:
Rename to have consistency with original plugin name https://github.com/WordPress/gutenberg/pull/50065#discussion_r1180827494

rowasc committed:
Rename to have consistency with original plugin name https://github.com/WordPress/gutenberg/pull/50065#discussion_r1180827494 (tests)

rowasc committed:
createInterpolateElement usages with CallExpression do not crash

Before, we had scenarios where mapping the first parameter's result for createInterpolateElement would crash due to not having a valid array of strings. This was caused because there are scenarios we do not currently support, but also were not bailing (instead we tried to parse them anyway, like calling a different method that is not guaranteed to return a string literal)

Now, we bail when we don't support a type of node, and parse the rest.

Added tests to identify success/errors when combining custom components w/createInterpolateElement.

rowasc committed:
Add documentation and reference original docs, too.

rowasc committed:
Added support for 'components' eslint/a11y rule options

https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md#rule-options

Now you can configure the linter in the same way as with the original, and use custom component wrappers that are taken as anchors

rowasc committed:
Docs should be .md

rowasc committed:
Remove all incidents of ignoring the linter for anchor-has-content

rowasc commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+const meta = {
+	type: 'problem',
+	docs: {
+		description:
```
```suggestion
		url: 'https://github.com/WordPress/gutenberg/tree/HEAD/packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md',
		description:
```

rowasc commented:
@gziolo thank you so much for the review, and I am sorry for the delay here, I got back from my meetup and finally got a chance to get back to this today.

> The old rule is disabled now, so removing all those comments from the code should. Do I miss something here?

No, not missing anything, sorry! I probably messed up the explanation, I will give it another go. Added the lint-ignore removal here now so that I can make sure they are all ✅ too. 

⚠️ I see the unit tests for PHP were failing in the previous commit, but it seems like it's unrelated and probably fine now - will monitor and come back if they fail again. 

@tyxla I would appreciate a review if you get a chance, especially after all the changes from today, but no rush at all. Other than that I will just follow your/all reviewers' direction as needed :) moving to "Ready to review" now that I added some docs.

rowasc committed:
Remove useless comment pointing to a removed dependency

tyxla commented on packages/eslint-plugin/configs/jsx-a11y.js:
```diff
@@ -8,6 +8,7 @@ module.exports = {
 				assert: 'htmlFor',
 			},
 		],
+		'jsx-a11y/anchor-has-content': 'off',
```
I think it's worth adding a comment here that we have a custom `@wordpress/jsx-a11y-anchor-has-content` rule. The comment could link to the rule README for example.

rowasc replied:
done! thanks! https://github.com/WordPress/gutenberg/pull/50065/files#diff-c46d7a1944225bc15703f31da819bf3948f250fb681696e1e5102fac0042f175R11

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+# jsx-a11y-anchor-has-content
+
+This rule wraps around [eslint-plugin-jsx-a11y/anchor-has-content](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md), continuing to enforce the original rule unless createInterpolateElement is being used to render an anchor element
```
I think we should change it to the rule name here.

Furthermore, I'm suggesting only minor drive-by formatting improvements.

```suggestion
This rule wraps around [jsx-a11y/anchor-has-content](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md), continuing to enforce the original rule unless `createInterpolateElement` is being used to render an anchor element.
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+# jsx-a11y-anchor-has-content
+
+This rule wraps around [eslint-plugin-jsx-a11y/anchor-has-content](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md), continuing to enforce the original rule unless createInterpolateElement is being used to render an anchor element
+In scenarios where an anchor is used through createInterpolateElement, this wrap-around rule identifies missing content in anchors for simple scenarios.
```
```suggestion
In scenarios where an anchor is used through `createInterpolateElement`, this wrap-around rule identifies missing content in anchors for simple scenarios.
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+
+## Rule details
+
+To use the optional object argument of type object (as in [eslint-plugin-jsx-a11y/anchor-has-content](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md)), configure it under the rule wrapper's name. See the [original documentation](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md) for other details.
```
```suggestion
To use the optional object argument of type object (as in [jsx-a11y/anchor-has-content](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md)), configure it under the rule wrapper's name. See the [original documentation](https://github.com/jsx-eslint/eslint-plugin-jsx-a11y/blob/main/docs/rules/anchor-has-content.md) for other details.
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+
+## Examples of correct usage
+
+With createInterpolateElement and a translate `__` function call:
```
```suggestion
With `createInterpolateElement` and a translate `__` function call:
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+With createInterpolateElement and a translate `__` function call:
+
+```jsx
+( ) => {
````
```suggestion
() => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+```jsx
+( ) => {
+				( <div>
+					   { createInterpolateElement( __( 'This is an <a> anchor with content </a>.' ), {
````
Minor, but I'd recommend not adding leading/trailing spaces as those can sometimes create unexpected visual issues, and examples are often copied from.

```suggestion
					   { createInterpolateElement( __( 'This is an <a>anchor with content</a>.' ), {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+		   };
+```
+
+With createInterpolateElement, without `__`:
````
```suggestion
With `createInterpolateElement`, without `__`:
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+With createInterpolateElement, without `__`:
+
+```jsx
+( ) => {
````
```suggestion
() => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+```jsx
+( ) => {
+				( <div>
+					   { createInterpolateElement( 'This is an <a> anchor with content </a>.', {
````
```suggestion
					   { createInterpolateElement( 'This is an <a>anchor with content</a>.', {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+		   };
+```
+
+Without createInterpolateElement:
````
```suggestion
Without `createInterpolateElement`:
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+```jsx
+( ) => {
+				( <div>
+					   This is an <a href="https://wordpress.org"> anchor with content </a>
````
```suggestion
					   This is an <a href="https://wordpress.org">anchor with content</a>
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+Without createInterpolateElement:
+
+```jsx
+( ) => {
````
```suggestion
() => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+		   };
+```
+
+With custom components configured (works with and without createInterpolateElement)
````
```suggestion
With custom components configured (works with and without `createInterpolateElement`)
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+
+```jsx
+// MyAnchor.js
+const MyAnchor = props => {
````
```suggestion
const MyAnchor = ( props ) => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+// MyAnchor.js
+const MyAnchor = props => {
+  return (
+    <a {...props}>{ props.children }</a>
```
Just minor coding standards suggestions here.

```suggestion
    <a { ...props }>{ props.children }</a>
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+// Something.js
+...
+return (
+  <MyAnchor>This is valid content</Anchor>
```
```suggestion
  <MyAnchor>This is valid content</MyAnchor>
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+
+## Examples of incorrect usage
+
+With createInterpolateElement and a translate `__` function call:
```
```suggestion
With `createInterpolateElement` and a translate `__` function call:
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+With createInterpolateElement and a translate `__` function call:
+
+```jsx
+( ) => {
````
```suggestion
() => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+		   };
+```
+
+With createInterpolateElement, without `__`:
````
```suggestion
With `createInterpolateElement`, without `__`:
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+With createInterpolateElement, without `__`:
+
+```jsx
+( ) => {
````
```suggestion
() => {
```

tyxla replied:
```suggestion
() => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+		   };
+```
+
+Without createInterpolateElement:
````
```suggestion
Without `createInterpolateElement`:
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+Without createInterpolateElement:
+
+```jsx
+( ) => {
````
```suggestion
() => {
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,134 @@
+		   };
+```
+
+With custom components configured (works with and without createInterpolateElement)
````
```suggestion
With custom components configured (works with and without `createInterpolateElement`)
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+  return (
+    <a {...props}>{ props.children }</a>
+  );
+}
```
```suggestion
const MyAnchor = ( props ) => {
  return (
    <a { ...props }>{ props.children }</a>
  );
}
```

tyxla commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,134 @@
+// Something.js - empty content
+...
+return (
+  <MyAnchor></Anchor>
```
```suggestion
  <MyAnchor></MyAnchor>
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Valid anchor with text content`,
+			code: `
+			( ) => {
```
```suggestion
			() => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+			code: `
+			( ) => {
+				( <div>
+					   { createInterpolateElement( __( 'This is a <a> link </a>.' ), {
```
```suggestion
					   { createInterpolateElement( __( 'This is a <a>link</a>.' ), {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Valid anchor with text content and an extra attribute`,
+			code: `
+			( ) => {
```
```suggestion
			() => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+			code: `
+			( ) => {
+				( <div>
+					   { createInterpolateElement( __( 'help me <a> test with multiple words </a>.' ), {
```
```suggestion
					   { createInterpolateElement( __( 'help me <a>test with multiple words</a>.' ), {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Valid anchor with a custom component as content.`,
+			code: `
+			( ) => {
```
```suggestion
			() => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+			code: `
+			( ) => {
+				( <div>
+					   { createInterpolateElement( __( 'help me <a> <TextWrapper /> </a>.' ), {
```
```suggestion
					   { createInterpolateElement( __( 'help me <a><TextWrapper /></a>.' ), {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Valid anchor without using createInterpolateElement`,
+			code: `
+			( ) => {
```
```suggestion
			() => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+			code: `
+			( ) => {
+				( <div>
+					   <a href="https://example.com"> test with multiple words </a>.
```
```suggestion
					   <a href="https://example.com">test with multiple words</a>.
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Invalid code, but out of scope for this rule (caught by jsx-a11y/anchor-has-content)`,
+			code: `
+			const a = ( ) => {
```
```suggestion
			const a = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Invalid usage of createInterpolateElement, ignored by the this rule`,
+			code: `
+			const a = ( ) => {
```
```suggestion
			const a = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+				createInterpolateElement,
+			} from '@wordpress/element';
+			import { __ } from '@wordpress/i18n';
+			const withTranslate = ( ) => {
```
```suggestion
			const withTranslate = () => {
```

tyxla replied:
Also, should this start with a capital letter? Otherwise, React [will complain](https://iq.js.org/questions/react/why-should-component-names-start-with-capital-letter).

Same for below usages of `<withTranslate />`

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+					</div>
+				)
+			}
+			const usesWithTranslate = ( ) => {
```
```suggestion
			const usesWithTranslate = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+				)
+			};
+
+			const withoutTranslateTwoLevels = ( ) => {
```
```suggestion
			const withoutTranslateTwoLevels = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Empty anchor content (@wordpress/jsx-a11y-anchor-has-content)`,
+			code: `
+			const a = ( ) => {
```
```suggestion
			const a = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Invalid anchor markup, with empty content (@wordpress/jsx-a11y-anchor-has-content)`,
+			code: `
+			const a = ( ) => {
```
```suggestion
			const a = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+		{
+			name: `Invalid anchor markup, with content (@wordpress/jsx-a11y-anchor-has-content)`,
+			code: `
+			const a = ( ) => {
```
```suggestion
			const a = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+				createInterpolateElement,
+			} from '@wordpress/element';
+			import { __ } from '@wordpress/i18n';
+			const failWithTranslate = ( ) => {
```
```suggestion
			const failWithTranslate = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+			const okUsesWithTranslate = ( ) => {
+				return (
+					<div>
+						{ createInterpolateElement( __( 'This one anchor <a> has content</a>. <failWithTranslate/>' ), {
```
```suggestion
						{ createInterpolateElement( __( 'This one anchor <a>has content</a>. <failWithTranslate/>' ), {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+					</div>
+				)
+			}
+			const okUsesWithTranslate = ( ) => {
```
```suggestion
			const okUsesWithTranslate = () => {
```

tyxla commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,332 @@
+				)
+			};
+
+			const failWithoutTranslateTwoLevels = ( ) => {
```
```suggestion
			const failWithoutTranslateTwoLevels = () => {
```

tyxla commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+	type: 'problem',
+	docs: {
+		description:
+			'Checks that anchors are not empty and have an href, relies on eslint-plugin-jsx-a11y/anchor-has-content under the hood. Supports usage of createInterpolateElement.',
```
```suggestion
			'Checks that anchors are not empty and have an href, relies on jsx-a11y/anchor-has-content under the hood. Supports usage of createInterpolateElement.',
```

tyxla commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+};
+
+/**
+ *
```
A description for each function could be useful. Applies to the following functions too.

rowasc replied:
done, thank!

tyxla commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+
+/**
+ *
+ * @param {*} tags     markup used in createInterpolateElement
+ * @param {*} tagNames names that we recognize as anchors, as configured in the eslint rules
+ * @return {Array} tagNames an array of the content of the <a> (or other) tags
```
We could vertically align the different sections for better readability. ESLint will usually complain, but then it is likely just not enabled here.

rowasc replied:
Thanks, agree. This is actually something I had to do because ESLint complained :( I had them aligned but it kept pestering me about it. Later I realized it was because the alignment rule is different if there is no description in the Doc block

rowasc replied:
https://github.com/WordPress/gutenberg/pull/50065#discussion_r1198223784 Apparently it also gets messed up in other cases - added more context in the link

tyxla commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+	const regex = new RegExp( `<(${ tagText })>(.*?)</(${ tagText })>`, 'g' );
+	const tagParts = [ ...tags.matchAll( regex ) ];
+	return tagParts.map( ( element ) => {
+		return element[ 2 ].trim();
```
Perhaps we actually intend to have the leading/trailing space?

rowasc replied:
yep, but in that case, it would not result in an empty anchor content (for the purposes of the rule, that's all that we care about) - however, the comment made me realize I can improve this abstraction a bit since it's not really clear what the intention is. Thanks!

As an example `<a> Hello world </a>` would be totally fine, because even if you remove trailing spaces you don't end up with an empty anchor, so the rule won't complain. However `<a> </a>` is not valid, as the result would be a '' after the trim, and the rule would complain at you.

tyxla commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+		// An unknown function call in a translate function may produce a valid string
+		// but verifying it is not straight-forward, so we bail
+		if ( args.filter( Boolean ).length === 0 ) {
+			nodeStr = null;
```
Why is this necessary if we override it immediately on L91?

rowasc replied:
😬 this was just a mistake while refactoring and I didn't have a test to cover it, thanks for catching it 
Added a quick scenario to cover it https://github.com/WordPress/gutenberg/pull/50065/files#diff-4c555e1042f88bdebf5cba43ba582ae78b721c7d64c1d0cc527ea1975ef191edR61

tyxla commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,161 @@
+			const componentOptions = options.components || [];
+			const typeCheck = [ 'a' ].concat( componentOptions );
+			if (
+				typeCheck.filter( ( type ) => type === node?.name?.name )
```
Do we also need to filter out nullish values before we check the `length`?

rowasc replied:
I didn't think so, but I've added a check to be safe in case something is misconfigured, thanks for catching it

tyxla reviewed (commented):
Lovely work, @rowasc! Really cool idea and nice work to wrap around the existing rule.

I've left a bunch of comments, but it's mostly code style and nothing major or blocking.

I think this already looks in a pretty stable state and the tests cover a great deal of the available scenarios. 🚀

tyxla reviewed (commented):
Lovely work, @rowasc! Really cool idea and nice work to wrap around the existing rule.

I've left a bunch of comments, but it's mostly code style and nothing major or blocking.

I think this already looks in a pretty stable state and the tests cover a great deal of the available scenarios. 🚀

gziolo commented:
> No, not missing anything, sorry! I probably messed up the explanation, I will give it another go. Added the lint-ignore removal here now so that I can make sure they are all ✅ too.

Excellent. Removing all comments that disable the original rule with all the unit tests covering possible cases is the best way to validate correctness and all we need to move forward 👍

For coding styles, you might want to try running ESLint with `--fix` flag as it might help with some comments that @tyxla left 😀

rowasc committed:
Update packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Add link to the new rule when disabling the anchor-has-content rule

rowasc committed:
Fix naming of JSX component which was not correct per https://iq.js.org/questions/react/why-should-component-names-start-with-capital-letter

rowasc committed:
Just add more context about the intention of the method

rowasc committed:
Anchor validation is moved to a (renamed) function instead of doing it in two steps

Before, we had a method that just returned the content of anchors and the actual validation was done by the main function in the linter. The problem there was that there wasn't a clear enough abstraction indicating the intent of the matcher that identified which tags were empty, causing confusion about it https://github.com/WordPress/gutenberg/pull/50065/files#r1187499276. Now, instead of getATagsContent we have validateAnchors which throws an error with the same code we report through eslint if anchors are empty or not valid.

Co-Authored-By: Boro <boro.sitnikovski@automattic.com>

rowasc committed:
- Return directly when we bail, the previous code would never do anything useful :)
- Fix issue where we were matching the wrong messageId for nodes with the latest refacotr

rowasc committed:
Add comment to the rule fn

rowasc committed:
Apply suggestions from code review

Adding all the code style edits from the reviewers' suggestions

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
if we ever misconfigure the custom component for eslint as falsy value, it won't be considered valid

rowasc committed:
Fix tests that called JSX with lowercase, add tests for expressions in translations

rowasc committed:
Rethrow if we have an unknown error during processing

Before we were automatically using the e.message thrown by the validator, which we had looked into and should not throw but to be safe, added a clause to rethrow if the unexpected thing happens instead of using the eslint report for normal non lint issues

rowasc committed:
Bring package-lock back to where it was in trunk

rowasc committed:
Apply suggestions from code review - docs' description section

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc committed:
Apply suggestions from code review

Co-authored-by: Marin Atanasov <8436925+tyxla@users.noreply.github.com>

rowasc commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,172 @@
+/**
+ *
+ * @file Wraps eslint-plugin-jsx-a11y/anchor-has-content to allow usage with createInterpolateElement.
```
```suggestion
 * @file Wraps around the rule jsx-a11y/anchor-has-content to allow usage with createInterpolateElement.
```

rowasc replied:
would it make sense to add a link to jsx-a11y/anchor-has-content in the file description too? or is this clear enough?

tyxla replied:
I think anyone can search for `jsx-a11y/anchor-has-content` and they'll find it, but you can also add a link to make it extra cool 😉

rowasc reviewed (commented):
Just responding to comments for now, I'll do a fresh read of the PR tomorrow with my last updates and ask for a review afterwards. 

thanks again all your help

rowasc committed:
Readability & description improvement for anchor validation function

rowasc committed:
Small readability improvements for getInterpolatedNodeText

rowasc committed:
Shortened sample code in rules docs

The previous samples were including more context than strictly necessary, making it harder to understand the point of the correct/incorrect usage scenarios. Hopefully this is minimal and still clear enough to be a good UX

rowasc committed:
Minimize footprint of test code while retaining quality

There was so much clutter that it was hard to understand the tests. We now get the same benefits with less code _shrugs_

rowasc committed:
Fix code style with --fix in docs

rowasc committed:
Test code style changed with --fix output for the more convoluted tests

rowasc commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
````diff
@@ -0,0 +1,104 @@
+}
+```
+
+## Examples of correct usage
````
After working a bit on making the code samples more succinct, I ran the ones I changed through `npm run lint:js myfile.js -- --fix` (file name is an example) so hopefully this will be ok now :)

rowasc commented on packages/eslint-plugin/rules/__tests__/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,278 @@
+/**
+ * External dependencies
+ */
+import { RuleTester } from 'eslint';
```
I made the test code a little less verbose since reading through it again made me realize it was hard to understand each test's content

rowasc committed:
Fix throws tag with {Error} - run --fix eslint

rowasc commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,173 @@
+ * Verify that the content of anchors is not empty, and that we find an anchor in the markup. The anchors' content
+ * can be from an <a> or other tags that are configured in the eslint 'component' rules (e.g. <MyAnchor>)
+ *
+ * @param {string} markup   markup used in createInterpolateElement
+ * @param {string} tagNames names that we recognize as anchors, as configured in the eslint rules
+ * @throws {Error}              an error if the markup does not contain a valid anchor or is empty
```
if I align this correctly, ESLint complains, *but* it is easy enough to solve it by configuring the jsdoc.js linter rules so that 'throws' is also counted as something that should be aligned. Happy to open a quick PR for that while this one is getting wrapped up, although the downside is that it'd cause a lot of new lint issues in the project and I'm not sure yet how y'all handle that (I could fix them, I'm just not sure if we'd want to do it so I want to check first) 

```

		'jsdoc/check-line-alignment': [
			'error',
			'always',
			{
				tags: [ 'param', 'arg', 'argument', 'property', 'prop', 'throws' ],
				preserveMainDescriptionPostDelimiter: true,
			},
		],
```

rowasc replied:
note: actually it wouldn't cause many  issues (I think there's just one other throw :) ) but it's still TBD if we want the change or not 

Changing the `@return` rule, on the other hand... would cause thousands of --fix changes.

rowasc reviewed (commented):
@gziolo  @tyxla thanks again for all the help - I updated the PR after feedback and after giving it another read. 

LMK if there's anything else I should change - I left a couple of suggestions that I was not sure about, too.

bor0 commented on packages/eslint-plugin/docs/rules/jsx-a11y-anchor-has-content.md:
```diff
@@ -0,0 +1,104 @@
+					"error", { "components": ["MyAnchor"] }
+				],
+    }
+}
```
```suggestion
{
	"rules": {
		"@wordpress/jsx-a11y-anchor-has-content": [
			"error", { "components": ["MyAnchor"] }
		],
	}
}
```

bor0 commented on packages/eslint-plugin/rules/jsx-a11y-anchor-has-content.js:
```diff
@@ -0,0 +1,173 @@
+			} catch ( e ) {
+				// if the error is not what we expect, something else went wrong and we should re-throw
+				if ( ! meta.messages.hasOwnProperty( e.message ) ) {
+					throw e;
```
The only other thing in `validateAnchorsContent` that can fail is the constructor of `RegExp`, but I'm wondering if we should really re-throw here, rather than report? Mainly curious how throwing compares to `context.report` in terms of developer experience, as throwing means that we completely stop the linting process.

I'm thinking something along the lines of:

```javascript
const meta = {
...
  messages: {
    unexpected: "Unexpected error occurred: {{error}}"
...
}
```

And then:

```suggestion
					context.report( {
						node: interpolatedNode,
						messageId: meta.messages.unexpected,
						data: { error: e.message },
					} );
```

Not sure if returning (silent failing) is another option, but we would probably want to report such errors.

rowasc replied:
Like you, I really don't think this should _ever_ happen, so with the re-throw, my intention was "If this ever happened, I'd prefer someone tells us about it rather than think it's a problem in their code.". 

The downside as we chatted about today is that yea, if it happens, it will stop the linter which is also not ideal. 

@tyxla do you have any thoughts on this decision? We could be more explicit in the report (and not rethrow) or add some hint about sending an issue, etc.

tyxla replied:
Yeah, I like Boro's idea 👍  I think if we just throw it might create a false expectation that the consumer has an issue in their code (or 3rd party code that they're using), while if we report it (with some additional context if possible), it can be the most helpful to them, bringing immediate value.

tyxla replied:
I do expect this shouldn't happen, but there are no downsides to having it handled just in case 😉

rowasc replied:
thank you! we just got it to report instead of throw and added a little nudge in the message :)

rowasc committed:
Do not throw if unexpected problems arise in our own code

Instead of re-throwing when there is a problem in the linter code, we use the eslint reporter to tell the user that this is not their fault :) we also give them context and nudge them to report it as a bug. This way they don't spend time chasing down something that is not their problem.

Co-Authored-By: Boro <boro.sitnikovski@automattic.com>

rowasc committed:
Trunk's package-lock (resetted it)
